### ISSUE TITLE
bulk action could not parse the date format corrrectly

### ISSUE THREAD

**Issue by foolcage** [elastic/elasticsearch-dsl-py#982]
I have some doc, when I save them one by one,it's well.

but when I add them to bulk,it shows:
```
[2018-08-07T20:25:31,827][DEBUG][o.e.a.b.TransportShardBulkAction] [eos_daily_user_statistic][0] failed to execute bulk item (index) BulkShardRequest [[eos_daily_user_statistic][0]] containing [index {[eos_daily_user_statistic][doc][b1_cryptocurrency_contract_RAM-EOS_2018-06-09], source[{"updateTimestamp": "20180609'T'115800.000Z", "turnoverIn": 1192.0, "volumeIn": 8192.0, "code": "RAM-EOS", "volume": 8192.0, "securityId": "cryptocurrency_contract_RAM-EOS", "turnoverOut": 0.0, "userId": "b1", "turnover": 1192.0, "volumeOut": 0.0, "name": "RAM/EOS", "timestamp": "2018-06-09", "id": "b1_cryptocurrency_contract_RAM-EOS_2018-06-09"}]}]
```
```
'reason': 'Invalid format: "20180609\'T\'115800.000Z" is malformed at "0609\'T\'115800.000Z"'
```
I did not set special date format and just use the default,I check:

```
basic_date_time
A basic formatter that combines a basic date and time, separated by a T: yyyyMMdd'T'HHmmss.SSSZ.
```

It seems the bulk action behavior is different comparing to one doc saving.

**Comment by HonzaKral** [comment]
Hi, can you please share what are your mappings and how are you creating your `bulk` actions? Thank you!

**Comment by foolcage** [comment]
@HonzaKral thanks for you reply at first

```
{
  "eos_user_statistic" : {
    "mappings" : {
      "doc" : {
        "dynamic" : "strict",
        "_all" : {
          "enabled" : false
        },
        "properties" : {
          "averagePrice" : {
            "type" : "float"
          },
          "cash" : {
            "type" : "float"
          },
          "code" : {
            "type" : "keyword"
          },
          "id" : {
            "type" : "keyword"
          },
          "name" : {
            "type" : "keyword"
          },
          "securityId" : {
            "type" : "keyword"
          },
          "timestamp" : {
            "type" : "date"
          },
          "turnover" : {
            "type" : "float"
          },
          "turnoverIn" : {
            "type" : "float"
          },
          "turnoverOut" : {
            "type" : "float"
          },
          "updateTimestamp" : {
            "type" : "date"
          },
          "userId" : {
            "type" : "keyword"
          },
          "volume" : {
            "type" : "float"
          },
          "volumeIn" : {
            "type" : "float"
          },
          "volumeOut" : {
            "type" : "float"
          }
        }
      }
    }
  }
}

```

the bulks is:
```
[
  {
    '_type': 'doc',
    '_source': {
      'timestamp': '2018-06-09',
      'turnoverOut': 0.0,
      'securityId': 'cryptocurrency_contract_RAM-EOS',
      'turnoverIn': 1201.0,
      'volumeOut': 0.0,
      'updateTimestamp': "20180609'T'121600.000Z",
      'id': 'g44tmobqhege_cryptocurrency_contract_RAM-EOS_2018-06-09',
      'volumeIn': 8192.0,
      'volume': 8192.0,
      'name': 'RAM/EOS',
      'turnover': 1201.0,
      'userId': 'g44tmobqhege',
      'code': 'RAM-EOS'
    },
    '_index': 'eos_daily_user_statistic',
    '_id': 'g44tmobqhege_cryptocurrency_contract_RAM-EOS_2018-06-09'
  },
  {
    '_type': 'doc',
    '_source': {
      'timestamp': '2018-06-09',
      'turnoverOut': 0.0,
      'securityId': 'cryptocurrency_contract_RAM-EOS',
      'turnoverIn': 1201.0,
      'volumeOut': 0.0,
      'updateTimestamp': "20180609'T'121600.000Z",
      'id': 'g44tmobqhege_cryptocurrency_contract_RAM-EOS',
      'volumeIn': 8192.0,
      'volume': 8192.0,
      'name': 'RAM/EOS',
      'turnover': 1201.0,
      'userId': 'g44tmobqhege',
      'code': 'RAM-EOS'
    },
    '_index': 'eos_user_statistic',
    '_id': 'g44tmobqhege_cryptocurrency_contract_RAM-EOS'
  },
  {
    '_type': 'doc',
    '_source': {
      'timestamp': '2018-06-09',
      'turnoverOut': 0.0,
      'securityId': 'cryptocurrency_contract_RAM-EOS',
      'turnoverIn': 1201.0,
      'volumeOut': 0.0,
      'updateTimestamp': "20180609'T'121600.000Z",
      'id': 'g44tmobrgige_cryptocurrency_contract_RAM-EOS_2018-06-09',
      'volumeIn': 8192.0,
      'volume': 8192.0,
      'name': 'RAM/EOS',
      'turnover': 1201.0,
      'userId': 'g44tmobrgige',
      'code': 'RAM-EOS'
    },
    '_index': 'eos_daily_user_statistic',
    '_id': 'g44tmobrgige_cryptocurrency_contract_RAM-EOS_2018-06-09'
  },
  {
    '_type': 'doc',
    '_source': {
      'timestamp': '2018-06-09',
      'turnoverOut': 0.0,
      'securityId': 'cryptocurrency_contract_RAM-EOS',
      'turnoverIn': 1201.0,
      'volumeOut': 0.0,
      'updateTimestamp': "20180609'T'121600.000Z",
      'id': 'g44tmobrgige_cryptocurrency_contract_RAM-EOS',
      'volumeIn': 8192.0,
      'volume': 8192.0,
      'name': 'RAM/EOS',
      'turnover': 1201.0,
      'userId': 'g44tmobrgige',
      'code': 'RAM-EOS'
    },
    '_index': 'eos_user_statistic',
    '_id': 'g44tmobrgige_cryptocurrency_contract_RAM-EOS'
  },
  {
    '_type': 'doc',
    '_source': {
      'timestamp': '2018-06-09',
      'turnoverOut': 0.0,
      'securityId': 'cryptocurrency_contract_RAM-EOS',
      'turnoverIn': 1201.0,
      'volumeOut': 0.0,
      'updateTimestamp': "20180609'T'121600.000Z",
      'id': 'g44tmobrhage_cryptocurrency_contract_RAM-EOS_2018-06-09',
      'volumeIn': 8192.0,
      'volume': 8192.0,
      'name': 'RAM/EOS',
      'turnover': 1201.0,
      'userId': 'g44tmobrhage',
      'code': 'RAM-EOS'
    },
    '_index': 'eos_daily_user_statistic',
    '_id': 'g44tmobrhage_cryptocurrency_contract_RAM-EOS_2018-06-09'
  },
  {
    '_type': 'doc',
    '_source': {
      'timestamp': '2018-06-09',
      'turnoverOut': 0.0,
      'securityId': 'cryptocurrency_contract_RAM-EOS',
      'turnoverIn': 1201.0,
      'volumeOut': 0.0,
      'updateTimestamp': "20180609'T'121600.000Z",
      'id': 'g44tmobrhage_cryptocurrency_contract_RAM-EOS',
      'volumeIn': 8192.0,
      'volume': 8192.0,
      'name': 'RAM/EOS',
      'turnover': 1201.0,
      'userId': 'g44tmobrhage',
      'code': 'RAM-EOS'
    },
    '_index': 'eos_user_statistic',
    '_id': 'g44tmobrhage_cryptocurrency_contract_RAM-EOS'
  },
  {
    '_type': 'doc',
    '_source': {
      'timestamp': '2018-06-09',
      'turnoverOut': 0.0,
      'securityId': 'cryptocurrency_contract_RAM-EOS',
      'turnoverIn': 1201.0,
      'volumeOut': 0.0,
      'updateTimestamp': "20180609'T'121600.000Z",
      'id': 'g44tmobrhege_cryptocurrency_contract_RAM-EOS_2018-06-09',
      'volumeIn': 8192.0,
      'volume': 8192.0,
      'name': 'RAM/EOS',
      'turnover': 1201.0,
      'userId': 'g44tmobrhege',
      'code': 'RAM-EOS'
    },
    '_index': 'eos_daily_user_statistic',
    '_id': 'g44tmobrhege_cryptocurrency_contract_RAM-EOS_2018-06-09'
  },
  {
    '_type': 'doc',
    '_source': {
      'timestamp': '2018-06-09',
      'turnoverOut': 0.0,
      'securityId': 'cryptocurrency_contract_RAM-EOS',
      'turnoverIn': 1201.0,
      'volumeOut': 0.0,
      'updateTimestamp': "20180609'T'121600.000Z",
      'id': 'g44tmobrhege_cryptocurrency_contract_RAM-EOS',
      'volumeIn': 8192.0,
      'volume': 8192.0,
      'name': 'RAM/EOS',
      'turnover': 1201.0,
      'userId': 'g44tmobrhege',
      'code': 'RAM-EOS'
    },
    '_index': 'eos_user_statistic',
    '_id': 'g44tmobrhege_cryptocurrency_contract_RAM-EOS'
  },
  {
    '_type': 'doc',
    '_source': {
      'timestamp': '2018-06-09',
      'turnoverOut': 0.0,
      'securityId': 'cryptocurrency_contract_RAM-EOS',
      'turnoverIn': 1201.0,
      'volumeOut': 0.0,
      'updateTimestamp': "20180609'T'121600.000Z",
      'id': 'g44tmobsgage_cryptocurrency_contract_RAM-EOS_2018-06-09',
      'volumeIn': 8192.0,
      'volume': 8192.0,
      'name': 'RAM/EOS',
      'turnover': 1201.0,
      'userId': 'g44tmobsgage',
      'code': 'RAM-EOS'
    },
    '_index': 'eos_daily_user_statistic',
    '_id': 'g44tmobsgage_cryptocurrency_contract_RAM-EOS_2018-06-09'
  },
  {
    '_type': 'doc',
    '_source': {
      'timestamp': '2018-06-09',
      'turnoverOut': 0.0,
      'securityId': 'cryptocurrency_contract_RAM-EOS',
      'turnoverIn': 1201.0,
      'volumeOut': 0.0,
      'updateTimestamp': "20180609'T'121600.000Z",
      'id': 'g44tmobsgage_cryptocurrency_contract_RAM-EOS',
      'volumeIn': 8192.0,
      'volume': 8192.0,
      'name': 'RAM/EOS',
      'turnover': 1201.0,
      'userId': 'g44tmobsgage',
      'code': 'RAM-EOS'
    },
    '_index': 'eos_user_statistic',
    '_id': 'g44tmobsgage_cryptocurrency_contract_RAM-EOS'
  }
]

```

The code:
```
            if self.es_actions:
                resp = elasticsearch.helpers.bulk(es_client, self.es_actions)
                self.logger.info("index success:{} failed:{}".format(resp[0], len(resp[1])))
                if resp[1]:
                    self.logger.error("error:{}".format(resp[1]))

```

And the error msg:
```
elasticsearch.helpers.BulkIndexError: ('500 document(s) failed to index.', [{'index': {'_type': 'doc', '_index': 'eos_daily_user_statistic', 'error': {'reason': 'failed to parse [updateTimestamp]', 'caused_by': {'reason': 'Invalid format: "20180609\'T\'121600.000Z" is malformed at "0609\'T\'121600.000Z"', 'type': 'illegal_argument_exception'}

```

**Comment by HonzaKral** [comment]
Problem is that `updateTimestamp` is just mapped as `date` without specific `format` which means it only support ISO formatted datetime or a unix timestamp in ms (as a long). If you want elasticsearch to support your custom format you need to specify it in the mappings.

Hope this helps!

**Comment by foolcage** [comment]
But with the same data in DocType, it could save successfully..

And from https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-date-format.html#built-in-date-formats

I shows
```
basic_date_time
A basic formatter that combines a basic date and time, separated by a T: yyyyMMdd'T'HHmmss.SSSZ.
```

I just use the format like this:
```
yyyyMMdd'T'HHmmss.SSSZ
```

**Comment by HonzaKral** [comment]
`basic_date_time` is one of the possible values for `format` which you can specify. The default value is `"strict_date_optional_time||epoch_millis"` (0)

when using `save()` what is the code? Are you by any chance using a python `datetime` object then? Because that we serialize to `isoformat` for you.

0 - https://www.elastic.co/guide/en/elasticsearch/reference/current/date.html

**Comment by foolcage** [comment]
the doc save successfully:
```
{
  '_type': 'doc',
  '_source': {
    'id': 'g44tmobqhege_cryptocurrency_contract_RAM-EOS_2018-06-09',
    'volume': 8192.0,
    'turnoverOut': 0.0,
    'name': 'RAM/EOS',
    'turnover': 1201.0,
    'securityId': 'cryptocurrency_contract_RAM-EOS',
    'timestamp': "20180609'T'000000.000Z",
    'code': 'RAM-EOS',
    'updateTimestamp': "20180609'T'121600.000Z",
    'volumeOut': 0.0,
    'turnoverIn': 1201.0,
    'volumeIn': 8192.0,
    'userId': 'g44tmobqhege'
  },
  '_id': 'g44tmobqhege_cryptocurrency_contract_RAM-EOS_2018-06-09',
  '_index': 'eos_user_statistic'
}
```

I check the save() code and it should just call doc.to_dict() to get the post body.

And I could confirm the dict just contain str timestamp,no datetime.

**Comment by foolcage** [comment]
And if  I want to set the special time format in the dsl Date field, is it possible?

thanks again for your reply.

**Comment by foolcage** [comment]
so why the same format,e.g  20180609'T'121600.000Z could save successfully using single doc but not in bulks? thank you.

**Comment by HonzaKral** [comment]
You can see in your document that a different value was used for the `save` action as it is in the proper iso format: `"updateTimestamp": "2018-06-09T12:16:00+00:00"`

Unfortunately the dsl right now doesn't support alternative formats in a general sense, but since your format is still supported by `dateutil` it should still work, to specify your format just pass a `format` kwarg to the `Date` field when defining the mapping, I would recommend something like: 
`Date(format="basic_date_time||strict_date_optional_time||epoch_millis")`

**Comment by foolcage** [comment]
In fact it's 20180609'T'121600.000Z ,the 2018-06-09T12:16:00+00:00 is got from es after saving.

Anyway I set the format in Date now, so what exactly the format string should be:
I try
```
format="yyyyMMdd'T'HHmmss.SSS||yyyy-MM-dd||epoch_millis"
```
the mapping could be saved well,it works just in  single doc saving not bulks.

```
format="yyyyMMddTHHmmss.SSS||yyyy-MM-dd||epoch_millis"
```
the mapping could not be saved.

After all,I set
```
timestamp = Date(format="yyyyMMdd HHmmss.SSS||yyyy-MM-dd||epoch_millis")
```
And just put ""20180609 121600.000",it works both single doc saving or bulks.

So what exactly the basic_date_time should be? just using the string 'basic_date_time' but not the format pattern?
